File: internal/agent/schema_test.go
package agent

import (
	"encoding/json"
	"testing"
)

func TestShortFormSchemaValidation(t *testing.T) {
	schema, err := CompileOutputSchema(json.RawMessage(`{"name":"string","score":"number?","confidence":"low|medium|high","active":"boolean"}`))
	if err != nil {
		t.Fatal(err)
	}
	valid := map[string]any{"name": "Acme", "score": nil, "confidence": "high", "active": true}
	if err := schema.Validate(valid); err != nil {
		t.Fatalf("expected valid output: %v", err)
	}
	invalid := map[string]any{"name": "Acme", "score": "wrong", "confidence": "certain", "active": true}
	if err := schema.Validate(invalid); err == nil {
		t.Fatal("expected invalid output to fail")
	}
}

func TestJSONSchemaValidation(t *testing.T) {
	raw := json.RawMessage(`{"type":"object","properties":{"count":{"type":"integer","minimum":1}},"required":["count"],"additionalProperties":false}`)
	schema, err := CompileOutputSchema(raw)
	if err != nil {
		t.Fatal(err)
	}
	if err := schema.Validate(map[string]any{"count": float64(2)}); err != nil {
		t.Fatalf("expected valid output: %v", err)
	}
	if err := schema.Validate(map[string]any{"count": float64(0)}); err == nil {
		t.Fatal("expected minimum violation")
	}
}

func TestRemoteSchemaReferenceRejected(t *testing.T) {
	_, err := CompileOutputSchema(json.RawMessage(`{"$ref":"https://example.com/schema.json"}`))
	if err == nil {
		t.Fatal("expected remote reference to be rejected")
	}
}

File: cloudflare-native/src/schema.test.ts
import { describe, expect, it } from "vitest";
import { compileOutputSchema } from "./schema";

describe("compileOutputSchema", () => {
  it("compiles and validates a standard JSON Schema", () => {
    const schema = compileOutputSchema({
      type: "object",
      properties: {
        name: { type: "string" },
        score: { type: ["number", "null"] },
        website: { type: "string", format: "uri" },
      },
      required: ["name", "score", "website"],
      additionalProperties: false,
    });

    expect(schema.validate({ name: "Acme", score: 9, website: "https://example.com" })).toBe(true);
    expect(schema.validate({ name: "Acme", score: null, website: "https://example.com" })).toBe(true);
    expect(schema.validate({ name: "Acme", score: "high", website: "not a URL" })).toBe(false);
  });

  it("rejects remote references", () => {
    expect(() =>
      compileOutputSchema({
        $ref: "https://example.com/schema.json",
      }),
    ).toThrow("Remote schema reference is not supported");
  });
});
